src/core/hle/service/ldr/ldr.cpp

```diff
@@ -31,35 +39,241 @@ constexpr ResultCode ERROR_NOT_INITIALIZED{ErrorModule::Loader, 87};
+
+    header.acid_fac_size = acid_file_access.size();
+    header.aci0_fac_size = aci0_file_access.size();
+    header.acid_sac_size = acid_service_access.size();
+    header.aci0_sac_size = aci0_service_access.size();
+
+    std::vector<u8> out(sizeof(ProgramInfoHeader) + acid_file_access.size() +
+                        aci0_file_access.size() + acid_service_access.size() +
+                        aci0_service_access.size());
+
+    u64 offset = 0;
+    std::memcpy(out.data() + offset, &header, sizeof(ProgramInfoHeader));
+    offset += sizeof(ProgramInfoHeader);
+    std::memcpy(out.data() + offset, acid_service_access.data(), acid_service_access.size());
+    offset += acid_service_access.size();
+    std::memcpy(out.data() + offset, aci0_service_access.data(), aci0_service_access.size());
+    offset += aci0_service_access.size();
+    std::memcpy(out.data() + offset, acid_file_access.data(), acid_file_access.size());
+    offset += acid_file_access.size();
+    std::memcpy(out.data() + offset, aci0_file_access.data(), aci0_file_access.size());
+
+    return out;
+}
+
+} // Anonymous namespace
+
 class DebugMonitor final : public ServiceFramework<DebugMonitor> {
 public:
-    explicit DebugMonitor() : ServiceFramework{"ldr:dmnt"} {
+    explicit DebugMonitor(const std::vector<Kernel::SharedPtr<Kernel::Process>>& process_list)
+        : ServiceFramework{"ldr:dmnt"}, process_list(process_list) {
         // clang-format off
         static const FunctionInfo functions[] = {
             {0, nullptr, "AddProcessToDebugLaunchQueue"},
             {1, nullptr, "ClearDebugLaunchQueue"},
-            {2, nullptr, "GetNsoInfos"},
+            {2, &DebugMonitor::GetNsoInfos, "GetNsoInfos"},
         };
         // clang-format on
 
         RegisterHandlers(functions);
     }
+
+private:
+    void GetNsoInfos(Kernel::HLERequestContext& ctx) {
+        IPC::RequestParser rp{ctx};
+        const auto process_id = rp.PopRaw<u64>();
+
+        auto write_size = ctx.GetWriteBufferSize() / sizeof(NSOInfo);
+
+        LOG_DEBUG(Service_LDR, "called, process_id={:016X}, write_size={:08X}", process_id,
+                  write_size);
+
+        const auto iter = std::find_if(
+            process_list.begin(), process_list.end(),
+            [process_id](const auto& process) { return process->GetProcessID() == process_id; });
+
+        if (iter == process_list.end()) {
+            LOG_ERROR(Service_LDR, "Cannot find process with given process ID!");
+            IPC::ResponseBuilder rb{ctx, 2};
+            rb.Push(ERROR_TITLE_NOT_FOUND);
+            return;
+        }
+
+        const auto& modules = (*iter)->GetModules();
+        const auto nso_infos = GetNSOInfo(modules);
+
+        write_size = std::min<std::size_t>(write_size, nso_infos.size());
+
+        ctx.WriteBuffer(nso_infos.data(), write_size * sizeof(NSOInfo));
+
+        IPC::ResponseBuilder rb{ctx, 3};
+        rb.Push(RESULT_SUCCESS);
+        rb.Push<u32>(write_size);
```

Comment:
```suggestion
        rb.Push(static_cast<u32>(write_size));
```

otherwise this is doing an implicit truncation